### PR TITLE
Upgrade OZ contracts with ERC721 approval workaround

### PR DESCRIPTION
**Pull request by TheMaxKim**
Tested trading:

ERC721: https://testnets.opensea.io/assets/0x766f9cd1ca9bc7444862bc69ecf8b1bddc462004/9
ERC1155: https://testnets.opensea.io/assets/0x11a1aba7c6c3e458325577cb74d5a59f6a69639b/4

Upgrading OZ contracts caused trading to fail due to `_isApprovedForOwner` not calling the appropriate overridden `isApprovedForAll` method. Fixed but not yet in the latest OZ release: https://github.com/OpenZeppelin/openzeppelin-contracts/pull/2597

### CODE REVIEW

**Review by sunwrobert — APPROVED** [review]

File: migrations/2_deploy_contracts.js [hunk 1]
```diff
@@ -72,10 +72,6 @@ module.exports = async (deployer, network, addresses) => {
     );
     const accessories = await CreatureAccessory.deployed();
     const factory = await CreatureAccessoryFactory.deployed();
-    await accessories.setApprovalForAll(
```
Comment (alexanderatallah) [inline migrations/2_deploy_contracts.js:75]: why is this no longer needed?

File: test/ERC1155Tradable.js [hunk 2]
```diff
@@ -300,8 +301,7 @@ contract("ERC1155Tradable - ERC 1155", (accounts) => {
              "0x0",
              {from: creator}
            ),
-           truffleAssert.ErrorType.revert,
-           'OVERFLOW'
```
Comment (alexanderatallah) [inline test/ERC1155Tradable.js:304]: can we include a string that should be in the error message?

File: contracts/ERC1155Tradable.sol [hunk 3]
```diff
@@ -41,33 +44,31 @@ contract ERC1155Tradable is ERC1155, ERC1155MintBurn, ERC1155Metadata, Ownable {
    * @dev Require msg.sender to own more than 0 of the token id
    */
   modifier ownersOnly(uint256 _id) {
-    require(balances[msg.sender][_id] > 0, "ERC1155Tradable#ownersOnly: ONLY_OWNERS_ALLOWED");
+    require(balanceOf(msg.sender, _id) > 0, "ERC1155Tradable#ownersOnly: ONLY_OWNERS_ALLOWED");
```
Comment (alexanderatallah) [inline contracts/ERC1155Tradable.sol:47]: should this be `_msgSender()`?

**Review by alexanderatallah — CHANGES_REQUESTED** [review]
added comments

**Comment by MINVEST01** [comment]
Hello you can include

  
>Вторник, 20 апреля 2021, 16:40 +03:00 от Alex Atallah ***@***.***>:
> 
> 
***@***.*** commented on this pull request.
>----------------------------------------------------------------------
>In  test/ERC1155Tradable.js :
>> @@ -300,8 +301,7 @@ contract("ERC1155Tradable - ERC 1155", (accounts) => {
>              "0x0",
>              {from: creator}
>            ),
>-           truffleAssert.ErrorType.revert,
>-           'OVERFLOW'
>
>can we include a string that should be in the error message?
>—
>You are receiving this because you commented.
>Reply to this email directly,  view it on GitHub , or  unsubscribe . 
 
 
С уважением к вашему делу

File: contracts/ERC721Tradable.sol [hunk 4]
```diff
@@ -14,8 +18,9 @@ contract ProxyRegistry {
  * @title ERC721Tradable
  * ERC721Tradable - ERC721 contract that whitelists a trading address, and has minting functionality.
  */
-contract ERC721Tradable is ERC721Full, Ownable {
+abstract contract ERC721Tradable is ERC721Enumerable, Ownable {
     using Strings for string;
```
Comment (flockonus) [inline contracts/ERC721Tradable.sol:22]: (my 2c as a bypasser) since you're using `abi.encodePacked` it seems safe to remove this line

Comment (flockonus) [reply contracts/ERC721Tradable.sol:22]: Really appreciate this PR btw

Comment (TheMaxKim) [reply contracts/ERC721Tradable.sol]: I believe you're right, removed it, nice catch!

File: test/ERC1155Tradable.js (hunk 2 above)
Comment (TheMaxKim) [reply test/ERC1155Tradable.js:304]: The error message only returns "revert". It's not that helpful, but I've added a check for it.

File: contracts/ERC1155Tradable.sol (hunk 3 above)
Comment (TheMaxKim) [reply contracts/ERC1155Tradable.sol]: Yes, these should be, updated.

File: migrations/2_deploy_contracts.js (hunk 1 above)
Comment (TheMaxKim) [reply migrations/2_deploy_contracts.js:75]: So I believe this actually wasn't doing anything meaningful given the setApprovalForAll ABI: https://github.com/0xsequence/erc-1155/blob/master/src/contracts/tokens/ERC1155/ERC1155.sol#L160

This was causing an issue with the upgraded OZ contracts as they checked whether the operator was the same as the sender: https://github.com/OpenZeppelin/openzeppelin-contracts/blob/v4.0.0/contracts/token/ERC1155/ERC1155.sol#L106

**Comment by TheMaxKim** [comment]
> added comments

Sorry for the delayed update on this, responded and updated.

**Comment by Striffly** [comment]
Are these modifications still necessary since the 4.1 update of OpenZeppelin?

**Comment by TheMaxKim** [comment]
> 
> 
> Are the ERC721 approval workaround changes still necessary since release of OpenZeppelin 4.1 ?

Ah, great, didn't catch that they had just released 4.1.0 which fixes the issue in question. Updating this.

**Review by alexanderatallah — APPROVED** [review]

**Pull request merged**